VariantShredding.md

```diff
@@ -25,290 +25,316 @@
+                for (name, field) in typed_value
+            }
+
+            if value is not None:
+                # this is a partially shredded object
+                assert isinstance(value, VariantObject), "partially shredded value must be an object"
+                assert typed_value.keys().isdisjoint(value.keys()), "object keys must be disjoint"
+
+                # union the shredded fields and non-shredded fields
+                return VariantObject(metadata, object_fields).union(VariantObject(metadata, value))
+
+            else:
+                return VariantObject(metadata, object_fields)
+
+        elif isinstance(typed_value, list):
+            # this is a shredded array
+            assert value is None, "shredded array must not conflict with variant value"
+
+            elements = [
+                construct_variant(metadata, elem.value, elem.typed_value)
+                for elem in list(typed_value)
+            ]
+            return VariantArray(metadata, elements)
+
+        else:
+            # this is a shredded primitive
+            assert value is None, "shredded primitive must not conflict with variant value"
+
+            return primitive_to_variant(typed_value)
+
+    elif value is not None:
+        return Variant(metadata, value)
+
+    else:
+        # value is missing
+        return None
+
+def primitive_to_variant(typed_value: Any): Variant:
+    if isinstance(typed_value, int):
+        return VariantInteger(typed_value)
+    elif isinstance(typed_value, str):
+        return VariantString(typed_value)
+    ...
 ```
 
-# Backward and forward compatibility
 
-Shredding is an optional feature of Variant, and readers must continue to be able to read a group containing only a `value` and `metadata` field.
+## Backward and forward compatibility
 
-Any fields in the same group as `typed_value`/`variant_value` that start with `_` (underscore) can be ignored.
-This is intended to allow future backwards-compatible extensions.
-In particular, the field names `_metadata_key_paths` and any name starting with `_spark` are reserved, and should not be used by other implementations.
-Any extra field names that do not start with an underscore should be assumed to be backwards incompatible, and readers should fail when reading such a schema.
+Shredding is an optional feature of Variant, and readers must continue to be able to read a group containing only `value` and `metadata` fields.
 
 Engines without shredding support are not expected to be able to read Parquet files that use shredding.
-Since different files may contain conflicting schemas (e.g. a `typed_value` column with incompatible types in two files), it may not be possible to infer or specify a single schema that would allow all Parquet files for a table to be read.
+Different files may contain conflicting schemas.
+That is, files may contain different `typed_value` columns for the same Variant with incompatible types.
```

Comment:
This has interesting implications for tools that I think already exist in parquet-mr (I thought there was something that could concatenate parquet files).  Is the intent that tools like this would fail?

Reply:
For concatenating files with different shredded variant types, a quick hack is to just make them non-shredded.

Reply:
Agreed, this is one reason though that we probably want to ensure a reference implementation can understand the type

Reply:
Utilities for concatenating Parquet files still require the same physical schema. For instance, concatenating a list of required elements with a list of optional elements is not possible without rewriting because the repetition level of the list element must match the encoding. This would be no different.

Reply:
Makes sense.